### PR TITLE
vkd3d: Workaround DXR bugs in Witcher 3 implementation.

### LINKED ISSUE #1342: Witcher 3 v 4.00 - turning on RTX locks up the game.
Distro: Arch
Proton: bleeding edge beta from steam
NVidia driver: 525.60.11-3 with PROTON_ENABLE_NGX_UPDATER=1 set

Game works fine with RTX turned off. turning RTX on reliably produces following error visible in steam stdout/stderr output:
`winevulkan/virtual_queue_worker: queue submission failed with -4, treating as DEVICE_LOST.`
It seems to also produce a correlated dmesg entry:
`[  973.501169] NVRM: Xid (PCI:0000:29:00): 109, pid=2503, name=witcher3.exe, Ch 0000002b, errorString CTX SWITCH TIMEOUT, Info 0x1cc029`

With yesterday's Proton (without 2.8 vkd3d) it produced a different error I failed to capture verbatim, but it included a fencing timeout.

### LINKED ISSUE #1373: Witcher 3 RT shadows are broken
After walking for a couple of meters, this happens:

### PR DESCRIPTION
**Pull request by HansKristian-Work**
Fixes #1373
Fixes #1342

### CODE REVIEW

**Comment by K0bin** [comment]
It fixes #1373, right?

**Comment by HansKristian-Work** [comment]
> It fixes #1373, right?

yes

File: libs/vkd3d/command.c [hunk 1]
```diff
@@ -7016,6 +7017,36 @@ static void STDMETHODCALLTYPE d3d12_command_list_Dispatch(d3d12_command_list_ifa
...
+        vk_barrier.dstStageMask = VK_PIPELINE_STAGE_2_ALL_COMMANDS_BIT;
+        vk_barrier.dstAccessMask = VK_ACCESS_2_MEMORY_READ_BIT | VK_ACCESS_2_MEMORY_WRITE_BIT;
+
+        dep_info.sType = VK_STRUCTURE_TYPE_DEPENDENCY_INFO;
+        dep_info.memoryBarrierCount = 1;
+        dep_info.pMemoryBarriers = &vk_barrier;
+
+        VK_CALL(vkCmdPipelineBarrier2(list->cmd.vk_command_buffer, &dep_info));
+        VKD3D_BREADCRUMB_TAG("ForceBarrier");
+    }
```
Comment (doitsujin) [inline libs/vkd3d/command.c:7049]: i'd kind of prefer to have all this factored out so that we can also do this for ExecuteIndirect if needed.

Comment (HansKristian-Work) [reply libs/vkd3d/command.c:7049]: Good point.

**Review by doitsujin — APPROVED** [review]

**Comment by kostea92** [comment]
The Witcher 3 next-gen — RT freeze on NVIDIA/Linux (Xid 109): FIXED with extra_rtas_sync

  Launch options that fixed it:
  PROTON_ENABLE_NVAPI=1 VKD3D_CONFIG=extra_rtas_sync %command%

  SYMPTOM
  Enabling ray tracing (DX12) freezes the picture within seconds while audio keeps
  playing. Kernel log shows:
    NVRM: Xid (PCI:0000:01:00): 109, name=witcher3.exe, CTX SWITCH TIMEOUT
  Game must be killed from a terminal.

  SYSTEM
  RTX 3060 12GB / driver 610.43.03 (open module) / CachyOS kernel 7.1.4 / KDE Wayland
  GE-Proton11-1 (vkd3d-proton 1.1-5388-gbee300e4c)

  RULED OUT (all tested)
  - Mods: crashes with -disablemods too
  - Resolution: crashes at 1080p
  - GPU load / weak GPU: CRASHES IN THE MENU the instant the RT toggle is flipped
  - VRAM: only 3-6 GB of 12 in use
  - Clocks: nvidia-smi -lgc 210,1800 changed nothing
  - Proton version: GE freezes on a frame, Experimental gives black screen + audio
  - VKD3D_CONFIG=single_queue / no_upload_hvv / force_host_cached: no fix
  - VKD3D_DISABLE_EXTENSIONS=VK_KHR_present_wait: no fix

  ROOT CAUSE
  PROTON_LOG=1 VKD3D_DEBUG=warn showed, ~10s before VK_ERROR_DEVICE_LOST:
    vkd3d_va_map_place_acceleration_structure: Attempted to place non-TLAS on
    VA #14ba05a00 previously used by TLAS, marking mutated
  The game reuses TLAS memory without inserting barriers, so rays traverse an
  invalid BVH and stall the RT core -> GPU context switch timeout.
  extra_rtas_sync inserts the RTAS barriers the game omits.

  CAVEATS
  - Mitigation, not a cure: an Xid 109 can still fire, but the GPU now recovers
    and the game survives instead of hanging dead.
  - If hangs persist, try: VKD3D_CONFIG=extra_rtas_sync,single_queue
  - ALT-TAB still kills the session (swapchain never restores, black screen).
  - Ignore old guides recommending VKD3D_CONFIG=dxr,dxr11 — those options no
    longer exist in modern vkd3d-proton.

**Pull request merged**